### PR TITLE
fix immutable put/gets

### PR DESCRIPTION
**Pull request by mafintosh**
this should fix BEP44 when putting/getting keys in the public bittorrent dht which currently doesn't work.

the spec is vague on how putting of keys work but after reading the libtorrent source code and pinging some bittorrent devs on irc it turns out you should do a .get before doing a .put and echoing the token back (similar to how announce works).

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -317,14 +316,14 @@ DHT.prototype._put = function (opts, cb) {
   var isMutable = opts.k
   var hash = isMutable
     ? sha1(opts.salt ? Buffer.concat([ opts.salt, opts.k ]) : opts.k)
-    : sha1(opts.v)
+    : sha1(bencode.encode(opts.v))
```
Comment (maxogden) [inline client.js]: this was one of the things we didnt realize until the end that made it finally work (by re-reading the spec a couple of times)

File: client.js [hunk 2]
```diff
@@ -362,25 +361,43 @@ DHT.prototype._put = function (opts, cb) {
...
-      data.a.token = opts.token || self._generateToken(node.addr)
-      data.a.seq = opts.seq
-      data.a.sig = opts.sign(encodeSigData(opts))
-      data.a.k = opts.k
-      if (opts.salt) data.a.salt = opts.salt
-      if (opts.cas) data.a.cas = opts.cas
+      q: 'get'
+    })
+
+    function putOnGet (err, res) {
```
Comment (maxogden) [inline client.js]: in the bep44 spec they actually don't document this in the request/response payloads, but you have to do a get first to each node, and that node returns you a write token. the write tokens are required for both mutable and immutable

**Comment by feross** [comment]
This looks fine to me, but I'm not super familiar with the BEP44 code. @substack, what do you think?

**Comment by mafintosh** [comment]
I'm gonna merge this now as it's currently broken and we need it for a new thing we're building. I'll make sure to maintain this feature for a bit in case we find other issues

**Comment by mafintosh** [comment]
Released in 4.0.2

**Comment by feross** [comment]
Sounds good @mafintosh

**Pull request merged**